### PR TITLE
Pr pg to redshift

### PR DESCRIPTION
**Pull request by meli-lewis**
Added functions to:

- write table from Postgres to a CSV in a temp directory
- divide that CSV into N groups, where N is the number of slices in Redshift
- write each group to S3
- generate a manifest for all groups
- copy groups to Redshift
- TEST!

### CODE REVIEW

**Comment by jklukas** [comment]
Do we understand the test failures?

**Comment by jklukas** [comment]
Is there a plan to add documentation on usage of this new functionality in the README? In evalutating this PR, it would be helpful to see a sketch of the intended usage.

File: shiftmanager/mixins/postgres.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        Instantiation is delayed until the object is first used.
+        """
+        print("Connecting to %s..." % self.pg_host)
+        return psycopg2.connect(user=self.pg_user,
+                                host=self.pg_host,
+                                port=self.pg_port,
+                                database=self.pg_database,
+                                password=self.pg_password)
+
+    def execute_and_commit_single_statement(self, statement):
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Since these methods are going to be mixed in to the `Redshift` class, I'd argue the names here should all include `pg` somewhere to make it explicit that they're interacting with Postgres rather than the main Redshift database the instance is connected to.

Could this be `pg_execute_and_commit`? And then we could have `copy_pg_table_to_csv`, etc?

File: shiftmanager/mixins/postgres.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        formatted_statement = copy.format(table_name=table_name,
+                                          csv_file_path=csv_file_path)
+        self.execute_and_commit_single_statement(formatted_statement)
+
+        row_count_select = "SELECT COUNT(*) from {table_name};".format(
+            table_name=table_name)
+        with self.pg_connection as conn:
+            with conn.cursor() as cur:
+                cur.execute(row_count_select)
+                row_count = [r for r in cur][0][0]
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: This might be more readable as `row_count = cur.fetchone()[0]`

File: shiftmanager/mixins/postgres.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        `chunk_size` rows per chunk.
+
+        Parameters
+        ----------
+        csv_file_path: str
+            File path for the CSV written by Postgres
+        row_count: int
+            Number of rows in the CSV
+        chunks: int
+            Number of chunks to return
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Should we also have a `Returns` section here? I honestly don't know whether it's reasonable to omit that if your initial description of the function states what the function returns. I mostly bring this up because you do have a `Returns` in the previous method here.

File: shiftmanager/mixins/postgres.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        left_closed_boundary.append(row_count - 1)
+        right_closed_boundary = left_closed_boundary[1:]
+        final_boundary_index = len(right_closed_boundary) - 1
+
+        # We're going to allocate a large buffer for this -- let's read as fast
+        # as possible
+        chunk_lines = []
+        boundary_index = 0
+        boundary = right_closed_boundary[boundary_index]
+        with open(csv_file_path, "r", 1048576) as f:
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Is this integer literal equivalent to one MB? Can you give it a descriptive name like `one_megabyte = 1048576`?

File: shiftmanager/mixins/postgres.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+            for count, row in enumerate(f):
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: `count` doesn't seem right here. This makes me think that we're getting a count of objects in each row.

I'm used to seeing `i` in this context, but even better would be `line_number`.

File: shiftmanager/mixins/postgres.py (hunk 3 above)
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Oh right, this is a generator. We should consistently use "yield" in the documentation rather than "return".

File: shiftmanager/mixins/postgres.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                    if boundary_index != final_boundary_index:
+                        boundary_index += 1
+                        boundary = right_closed_boundary[boundary_index]
+                    yield "".join(chunk_lines)
+                    chunk_lines = []
+
+    def write_csv_chunk_to_s3(self, chunk, bucket, s3_key_path):
+        """
+        Given a string chunk that represents a piece of a CSV file, write
+        the chunk to a Boto S3 key.
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: "Boto S3 key" sounds odd. No need to specify "Boto" here.

File: shiftmanager/mixins/postgres.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+        Parameters
+        ----------
+        chunk: str
+            String blob representing a chunk of a larger CSV.
+        bucket: boto.s3.bucket.Bucket
+            The bucket we're writing to
+        s3_key_path: str
+            The key path to write the chunk to
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: I like as much parallel structure and consistency in docs as possible. Either end every parameter description with a period, or none of them. Either use "we're writing to" consistently or the more passive "to write to" consistently. I honestly am not sure what style is dominant in the documentation right now.

If you glance through the existing docs and we're already pretty fragmented in style, then you can ignore this and we'll try to address that separately in the future.

File: shiftmanager/mixins/postgres.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        """
+        boto_key = bucket.new_key(s3_key_path)
+        boto_key.set_contents_from_string(chunk, encrypt_key=True)
+
+    def generate_copy_statement(self, table_name, manifest_key_path):
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Seems like this should be named `get_copy_statement` to be consistent with other methods here. `generate` invites the possibility of confusion with Python generators.

File: shiftmanager/mixins/postgres.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+        """
+        Write the contents of a Postgres table to Redshift.
+        The table will be written to the given bucket under the given
+        key prefix.
+
+        Parameters
+        ----------
+        table_name: str
+            Table name to be written to CSV
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Are we assuming the same name for the source table in PG and the destination table in Redshift? Seems like those should be separate parameters.

File: shiftmanager/mixins/postgres.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        # All keys written to S3 in the event cleanup is needed
+        all_s3_keys = []
+
+        if not key_prefix.endswith("/"):
+            final_key_prefix = "".join([key_prefix, "/"])
+        else:
+            final_key_prefix = key_prefix
+
+        try:
+            fp, csv_temp_path = mkstemp()
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Is the intention here to create a directory? If so, I think this should be `mkdtemp` instead, which also avoids creating the unused file handle.

File: shiftmanager/mixins/s3.py [hunk 11]
```diff
@@ -37,6 +37,11 @@ def wrapper(self, *args, **kwargs):
 class S3Mixin(object):
     """The S3 interaction base class for `Redshift`."""
 
+    s3_conn = None
+    aws_access_key_id = None
+    aws_secret_access_key = None
+    security_token = None
```
Comment (jklukas) [inline shiftmanager/mixins/s3.py]: By putting these here, I think these are class-level attributes, so they're shared between different `Redshift` instances in the same program, which I don't think is what we want.

These should be handled as instance attributes accessed as `self.s3_conn`, etc. If we need to initialize them, that should happen in the constructor.

File: shiftmanager/mixins/postgres.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+from tempfile import mkstemp
+
+import psycopg2
+
+from shiftmanager.memoized_property import memoized_property
+from shiftmanager.mixins.s3 import S3Mixin
+from shiftmanager import util
+
+
+class PostgresMixin(S3Mixin):
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Some of the methods defined in this file don't depend on PG at all, but rather are helpers that interact with S3 that we end up using to support PG operations. Would it make sense to factor out those methods out to the `S3Mixin`?

**Comment by wrobstory** [comment]
> Do we understand the test failures?

 A number of these tests require a running Postgres instance to work. We [tagged those](https://github.com/SimpleFinance/shiftmanager/pull/7/files#diff-9d0c917b402bd139cb098a4b7e1658abR59) with `postgrestest` so that we can run py.test [with a flag to exclude those tests](http://pytest.org/latest/example/markers.html). 

BUT we haven't updated the Travis file to tell it to do so yet.

File: shiftmanager/mixins/postgres.py (hunk 1 above)
Comment (wrobstory) [inline shiftmanager/mixins/postgres.py]: Agreed.

File: shiftmanager/mixins/postgres.py (hunk 9 above)
Comment (wrobstory) [inline shiftmanager/mixins/postgres.py]: We are making this assumption, but this is a good idea!

File: shiftmanager/mixins/postgres.py (hunk 10 above)
Comment (wrobstory) [inline shiftmanager/mixins/postgres.py]: :+1:

File: shiftmanager/mixins/postgres.py (hunk 12 above)
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Also, I should explain a bit the intention behind the Mixins. None of them can actually stand on their own; they make assumptions about stuff in the main `Redshift` class (like the actual Redshift connection). They exist solely to put some more structure around the code and not have everything in a single class in a really big file.

So, that said, I don't know how I feel about dependencies between `Mixins`. Clearly, some of the stuff in this `PostgresMixin` relies on functionality from `S3Mixin`. Do we want to make that dependency explicit by inheriting from `S3Mixin` here? Or, should we just say it's fine to use any method from any mixin, because these are always going to be called from the context of a `Redshift` class which inherits from all the mixins anyway?

File: shiftmanager/redshift.py [hunk 13]
```diff
@@ -10,11 +10,11 @@
 
 import psycopg2
 
-from shiftmanager.mixins import AdminMixin, ReflectionMixin, S3Mixin
+from shiftmanager.mixins import (AdminMixin, ReflectionMixin, PostgresMixin)
 from shiftmanager.memoized_property import memoized_property
 
 
-class Redshift(AdminMixin, ReflectionMixin, S3Mixin):
+class Redshift(AdminMixin, ReflectionMixin, PostgresMixin):
```
Comment (jklukas) [inline shiftmanager/redshift.py]: Related to my previous comment about Mixins depending on each other, it feels weird that `S3Mixin` now doesn't appear as an explicit dependency of the `Redshift` class. I read this and think, _why did we skip the S3Mixin_?

And this also brings up an interesting question. What happens if the same class appears twice in the inheritance tree? It's probably fine; I'd imagine it's no different than inheriting from two different classes that have methods with the same name. There's a method resolution order that's followed, and one of the classes gets precedence based on the order of inheritance.

I'm tempted to say that `S3Mixin` should appear here whether or not `PostgresMixin` inherits from `S3Mixin` in the end.

Comment (jklukas) [inline shiftmanager/redshift.py]: Very interested in other thoughts on this.

File: shiftmanager/redshift.py [hunk 14]
```diff
@@ -98,3 +97,29 @@ def mogrify(self, batch, parameters=None, execute=False):
...
+        """
+        with self.connection as conn:
+            with conn.cursor() as curs:
+                curs.execute("""select distinct(tablename)
+                                from pg_table_def
+                                where tablename = '{}';""".format(table_name))
+
+                table = [row for row in curs]
+
+        if table and table[0][0] == table_name:
```
Comment (jklukas) [inline shiftmanager/redshift.py]: `fetchone` would be good here as well.

Comment (jklukas) [inline shiftmanager/redshift.py]: The query is either going to return one row or now rows, correct?

It might be more clear to make the query a `COUNT DISTINCT`, so it'll return one row with value of either 0 or 1. We can then do `table_count = curs.fetchone()[0]` and then we'd `return (table_count == 1)`.

File: shiftmanager/redshift.py [hunk 15]
```diff
@@ -98,3 +97,29 @@ def mogrify(self, batch, parameters=None, execute=False):
...
+            return True
+        else:
+            return False
```
Comment (jklukas) [inline shiftmanager/redshift.py]: It would be more idiomatic to avoid the if/then here and do `return (table and table[0][0] == table_name)`, or `return (table_count == 1)` as discussed in the previous comment.

File: shiftmanager/redshift.py [hunk 16]
```diff
@@ -98,3 +97,29 @@ def mogrify(self, batch, parameters=None, execute=False):
...
+        ----------
+        table_name : str
+            The name of the table for whose existence we're checking
+
+        Returns
+        -------
+        boolean
+        """
+        with self.connection as conn:
+            with conn.cursor() as curs:
```
Comment (jklukas) [inline shiftmanager/redshift.py]: I normally name this `cur` rather than `curs`. One could argue either way, but the existing code base, I believe, is consistent in using `cur`.

**Comment by jklukas** [comment]
I'm done commenting for this round, I think. I have not looked through the tests, but can do so in another session if you'd like feedback there.

**Comment by jklukas** [comment]
Taking a step back, let's discuss overall workflow for this table copy process.

This code assumes that you're loading into a Redshift table that has the same structure as the source table. Do we want to make that assumption? What if you wanted to load into a table with some extra columns not present in the source PG table? In such a case, you'd either have to do some custom transformation of the CSV files before loading, or you'd need to do the load into a staging table in Redshift, and then follow up with some custom `INSERT` statement to copy from the staging table into the final destination table.

I kind of like the second possibility here, where you'd be loading into a staging table in order to get the data into Redshift initially, then you have freedom to put it where you want from there. In such a workflow, it would be wonderfully convenient if we had an option for shiftmanager to *create* the staging table, copying the schema from Postgres and massaging it to a CREATE TABLE statement appropriate for Redshift. That's potentially a significant project in itself, though.

As we document this process, we should make it explicit what assumptions we're making about what preparation you need to do in Redshift beforehand to make this work.

File: shiftmanager/tests/test_postgres.py [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+    key_list = [x.split("_")[3] for x in bucket_keys]
+
+    comparison_list = ["".join([str(y), '.csv']) for y in range(0, 10)]
+
+    assert key_list == comparison_list
+
+    copy_statement = cur.statements[-1]
+
+    from pprint import pprint;import pytest;pytest.set_trace()
```
Comment (wrobstory) [inline shiftmanager/tests/test_postgres.py]: oops!

File: shiftmanager/redshift.py (hunk 13 above)
Comment (wrobstory) [inline shiftmanager/redshift.py]: > I'm tempted to say that S3Mixin should appear here whether or not PostgresMixin inherits from S3Mixin in the end.

Lots of learnings today about Python MRO. 

So this works: 
```python
class Redshift(AdminMixin, ReflectionMixin, PostgresMixin, S3Mixin):
```

But this does not:
```python
    class Redshift(AdminMixin, ReflectionMixin, S3Mixin, PostgresMixin):
E   TypeError: Cannot create a consistent method resolution
E   order (MRO) for bases object, PostgresMixin, S3Mixin
```
The reason it wasn't included here in the first place is because I never tried the first way- we only tried the latter, and it broke, so we moved on. 

I am definitely +1 on including it as `class Redshift(AdminMixin, ReflectionMixin, PostgresMixin, S3Mixin)`

File: shiftmanager/mixins/postgres.py (hunk 12 above)
Comment (wrobstory) [inline shiftmanager/mixins/postgres.py]: > Do we want to make that dependency explicit by inheriting from S3Mixin here?

Now that we've verified that we can [inherit from both](https://github.com/SimpleFinance/shiftmanager/pull/7/files#r51158621) with proper MRO, I think it's best to include it here to show that we have a hard dependency on it. 

> Some of the methods defined in this file don't depend on PG at all, but rather are helpers that interact with S3 that we end up using to support PG operations.

The only one that stands out to me is  `write_csv_chunk_to_s3`. I agree that we should move this to `S3Mixin` and call it something like `write_string_to_s3`, because fundamentally that's all it is doing.

File: shiftmanager/mixins/s3.py (hunk 11 above)
Comment (wrobstory) [inline shiftmanager/mixins/s3.py]: I agree, but am going to have to think about this a bit- right now no mixins have constructors.

Comment (jklukas) [inline shiftmanager/mixins/s3.py]: As long as the constructor doesn't need to take parameters (which it doesn't in this case), I don't think there's any harm in having a constructor here.

Except that it'll probably get called twice now, since S3Mixin shows up twice in the hierarchy.

File: shiftmanager/redshift.py (hunk 13 above)
Comment (jklukas) [inline shiftmanager/redshift.py]: Thanks for checking into this. It sounds reasonable to do `class Redshift(AdminMixin, ReflectionMixin, PostgresMixin, S3Mixin)` and also have `PostgresMixin` directly inherit from `S3Mixin`.

File: shiftmanager/tests/test_postgres.py (hunk 17 above)
Comment (jklukas) [inline shiftmanager/tests/test_postgres.py]: Bad rob.

**Comment by jklukas** [comment]
I think I've responded to all of @wrobstory's commentary.

File: shiftmanager/mixins/postgres.py [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+                  credentials 'aws_access_key_id={key_id};aws_secret_access_key={secret_key}'
+                  manifest
+                  csv;""".format(table_name=table_name,
+                                 manifest_key_path=manifest_key_path,
+                                 key_id=self.aws_access_key_id,
+                                 secret_key=self.aws_secret_access_key)
+
+    def copy_table_to_redshift(self, table_name, bucket_name, key_prefix,
+                               slices):
+
```
Comment (xstevens) [inline shiftmanager/mixins/postgres.py]: delete extra space between function and docstring

File: shiftmanager/mixins/s3.py [hunk 19]
```diff
@@ -110,7 +121,7 @@ def write_dict_to_key(self, data, key, close=False):
         return key
 
     @check_s3_connection
-    @memoize
+    # @memoize
```
Comment (xstevens) [inline shiftmanager/mixins/s3.py]: this is commented out? should we just remove it?

File: shiftmanager/mixins/postgres.py [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+        -------
+        row_count: int
+        """
+
+        copy = "COPY {table_name} TO '{csv_file_path}' DELIMITER ',' CSV;"
+        formatted_statement = copy.format(table_name=table_name,
+                                          csv_file_path=csv_file_path)
+        self.execute_and_commit_single_statement(formatted_statement)
+
+        row_count_select = "SELECT COUNT(*) from {table_name};".format(
```
Comment (xstevens) [inline shiftmanager/mixins/postgres.py]: How is this count used? And is it okay that the count may not match the number copied above if something else is actively inserting?

Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: > is it okay that the count may not match the number copied above if something else is actively inserting?

That's a fantastic question. It's probably fine to ignore any new rows since a backfill is going to have to represent a point in time anyway. But we should make sure this code handles that situation gracefully and doesn't make assumptions that would raise an exception or something.

File: shiftmanager/mixins/postgres.py (hunk 7 above)
Comment (meli-lewis) [inline shiftmanager/mixins/postgres.py]: https://github.com/SimpleFinance/shiftmanager/blob/master/shiftmanager/util.py shows inconsistency in parameter description punctuation, but has consistent tense. My preferences (none of which ought to violate [docstring conventions](https://www.python.org/dev/peps/pep-0257/) if I read them correctly):

Summary line ends with a period, as specified by 

> The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

Nothing is said in PEP257 or PEP8 about parameter descriptions, and all but one of mine were without punctuation, so I'll correct the one and proceed without punctuation hereafter.

I'll keep the passive voice.

Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Those all seem like reasonable choices. We should collaborate someday on writing a `pedantic-docstring-linter` :tm: to slavishly enforce rules like these.

File: shiftmanager/mixins/postgres.py (hunk 20 above)
Comment (wrobstory) [inline shiftmanager/mixins/postgres.py]: Agreed on fantastic question. 

I went to the docs and guess what:

> On successful completion, a COPY command returns a command tag of the form
   `COPY count`
   The count is the number of rows copied.

Neato -  so can we get this from psycopg2? YEP: 

```
(Pdb) cur.statusmessage
'COPY 300'
(Pdb) cur.rowcount
300
```

So, instead of making the `count` call, we can simple return `cur.rowcount`. 

SOHAPPY

Comment (wrobstory) [inline shiftmanager/mixins/postgres.py]: One more thing: I also realized/remembered that  `COPY` can take a `select` statement to determine which rows to copy. I think we probably want to add this, as it would make this a lot more flexible.

File: shiftmanager/redshift.py (hunk 14 above)
Comment (jklukas) [inline shiftmanager/redshift.py]: To be clear, I think the query would be:
```sql
SELECT COUNT(DISTINCT tablename)
FROM pg_table_def
WHERE tablename = '{}'
```

File: shiftmanager/mixins/postgres.py (hunk 9 above)
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: I'd still like to see the postgres table name broken out to a separate parameter here.

File: shiftmanager/mixins/postgres.py [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+        formatted_statement = copy.format(table_name=table_name,
+                                          csv_file_path=csv_file_path)
+        self.pg_execute_and_commit_single_statement(formatted_statement)
+
+        row_count_select = "SELECT COUNT(*) from {table_name};".format(
+            table_name=table_name)
+        with self.pg_connection as conn:
+            with conn.cursor() as cur:
+                cur.execute(row_count_select)
+                row_count = cur.fetchone()[0]
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: I think @wrobstory had suggested we can avoid this extra query and get a more reliable return value by checking the row count returned by the COPY statement itself.

File: shiftmanager/mixins/postgres.py [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+        return """copy {table_name}
+                  from '{manifest_key_path}'
+                  credentials 'aws_access_key_id={key_id};aws_secret_access_key={secret_key}'
+                  manifest
+                  csv;""".format(table_name=table_name,
+                                 manifest_key_path=manifest_key_path,
+                                 key_id=self.aws_access_key_id,
+                                 secret_key=self.aws_secret_access_key)
+
+    def copy_table_to_redshift(self, pg_table_name, redshift_table_name,
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Can this be `copy_pg_table_to_redshift` to make it explicit what the data source is?

File: shiftmanager/mixins/postgres.py [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+            for line_number, row in enumerate(f):
+                chunk_lines.append(row)
+                if line_number == boundary:
+                    if boundary_index != final_boundary_index:
+                        boundary_index += 1
+                        boundary = right_closed_boundary[boundary_index]
+                    yield "".join(chunk_lines)
+                    chunk_lines = []
+
+    def create_copy_statement(self, table_name, manifest_key_path):
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Can we prefix this name with an underscore, to indicate that it's not intended to be user-facing? This is creating a very particular copy statement with particular parameters, so it should have a more specific name if it's going to be part of the API.

File: shiftmanager/mixins/postgres.py [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+        # All keys written to S3 in the event cleanup is needed
+        all_s3_keys = []
+
+        if not key_prefix.endswith("/"):
+            final_key_prefix = "".join([key_prefix, "/"])
+        else:
+            final_key_prefix = key_prefix
+
+        try:
+            # csv_temp_path = mkdtemp()
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Commented line should be removed.

File: shiftmanager/mixins/postgres.py [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+            fp, csv_temp_path = mkstemp()
```
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: Instead of a `try` statement here where we call `mkstemp` and then clean up in the `finally`, this can be a `with tempfile.NamedTemporaryFile() as tf:`. Then `csv_temp_path = tf.name`.

The file will be deleted automatically at then end of the `with` context, so you don't need to write your own cleanup code.

**Comment by jklukas** [comment]
What's the status of Travis tests? Are there tests running right now that we intend to have Travis skip?

**Comment by wrobstory** [comment]
> What's the status of Travis tests? Are there tests running right now that we intend to have Travis skip?

Ok, we now have a passing built as of [this commit](https://github.com/SimpleFinance/shiftmanager/commit/a897552264db236163506cd8157a7cec5b2e8923).

**Comment by meli-lewis** [comment]
tumblr_mi8vsyefq71s2f6gpo1_250

File: shiftmanager/mixins/admin.py [hunk 26]
```diff
@@ -140,5 +140,5 @@ def alter_user(self, name,
                 options.append("RESET %s" % param)
             else:
                 options.append("SET %s = %s" % (param, value))
-        statement += ', '.join(options)
+        statement += ' '.join(options)
```
Comment (jklukas) [inline shiftmanager/mixins/admin.py]: We should make sure that #8 gets merged before this one, as this change is unrelated to the rest of what's going on in this PR.

File: shiftmanager/mixins/postgres.py (hunk 10 above)
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: @meli-lewis clarified that this is indeed intended to be a single temp file, but I added a comment that looks to be buried now that I'd prefer to see `tempfile.NamedTemporaryFile` used here in a `with` statement as a way of avoiding the explicit `try`/`finally` logic here.

File: shiftmanager/redshift.py (hunk 13 above)
Comment (jklukas) [inline shiftmanager/redshift.py]: Looks like this discussion is still unresolved. It looks like we have agreement that `class Redshift(AdminMixin, ReflectionMixin, PostgresMixin, S3Mixin)` should be preferred, so I'd like to see this line changed to match.

File: shiftmanager/redshift.py [hunk 27]
```diff
@@ -74,6 +73,8 @@ def __init__(self, database=None, user=None, password=None, host=None,
 
         self._all_privileges = None
 
+        super(Redshift, self).__init__()
```
Comment (jklukas) [inline shiftmanager/redshift.py]: What's the intention of this call. This gets into the details of method resolution order, which I can never remember and am not sure I totally understand.

My assumption is that this is going to resolve to the constructor of the first class in the list of classes `Redshift` inherits from, so `AdminMixin`.

I believe `S3Mixin` is the only mixin that contains a constructor, so that's the only one we care about calling. Is the intention here to call the `S3Mixin` constructor?

I think this is the reason why `self.s3_conn = None` occurred in the constructor for `Redshift` rather than in the `S3Mixin`. It mixes up responsibilities a bit, but it avoids this constructor problem.

Was initializing `s3_conn` here rather than in `S3Mixin` causing a problem? Or was this moved for philosophical  rather than practical reasons?

File: tox.ini [hunk 28]
```diff
@@ -35,4 +35,5 @@ addopts =
     --doctest-modules
     --ignore=setup.py
     --ignore=docs/conf.py
+    -m "not postgrestest"
```
Comment (jklukas) [inline tox.ini]: This is interesting. I was wondering how hiding tests was going to work with pytest.

**Comment by jklukas** [comment]
This is getting close! Getting through review on a big PR is always a bit of a process. I'm done responding to this round of changes.

File: shiftmanager/redshift.py (hunk 27 above)
Comment (wrobstory) [inline shiftmanager/redshift.py]: > Is the intention here to call the S3Mixin constructor

Yes.

> Was initializing s3_conn here rather than in S3Mixin causing a problem?

Yeah, wanted to be able to initialize the Postgres + S3 Mixins with one another without involving Redshift at all so we could isolate the Postgres tests from the rest of the test suite. 

Unfortunately, this ended up being the solution: https://github.com/SimpleFinance/shiftmanager/pull/7/files#diff-862ec20433c5ea312a4172f245bd0a7eR139

So I suppose given that, we could put `s3_conn` back into the `Redshift` constructor. However, I think it feels more correct to be in the `S3Mixin`, particularly given that this logic lives there: https://github.com/SimpleFinance/shiftmanager/blob/master/shiftmanager/mixins/s3.py#L21

Also, it does not require that any other classes that want to use `S3Mixin` have to know to have `s3_conn` in their constructors. 

This does raise the point that with a constructor, perhaps we should simply name it `S3` and not `S3Mixin`.

Comment (jklukas) [inline shiftmanager/redshift.py]: > with a constructor, perhaps we should simply name it S3 and not S3Mixin.

I don't see having a constructor as disqualifying this as a Mixin. We don't have any intention for this to be able to stand on its own outside of `Redshift`, so it's still a Mixin to me.

I think I'm convinced that it's worth keeping `s3_conn` initialization within the mixin rather than in `Redshift`. I see two potential ways forward:

* Drop the use of `super` and explicitly call `S3Mixin.__init__()` instead, *or*
* We could just not have a constructor at all. We expect that all methods using `s3_conn` are decorated with `@check_s3_connection`. We could add a check there to see if `self.s3_conn` exists and initialize there rather than in the constructor (I think this would be `if not self.hasattr("s3_conn"): self.s3_conn = None`). Then we'd never have to worry about subclasses remembering to call the S3Mixin constructor

File: shiftmanager/mixins/postgres.py (hunk 12 above)
Comment (jklukas) [inline shiftmanager/mixins/postgres.py]: This happened! :tada:

File: shiftmanager/redshift.py (hunk 27 above)
Comment (wrobstory) [inline shiftmanager/redshift.py]: > Drop the use of super and explicitly call S3Mixin.__init__() instead, or
We could just not have a constructor at all. We expect that all methods using s3_conn are decorated with @check_s3_connection. We could add a check there to see if self.s3_conn exists and initialize there rather than in the constructor (I think this would be if not self.hasattr("s3_conn"): self.s3_conn = None). Then we'd never have to worry about subclasses remembering to call the S3Mixin constructor

I would prefer the former. If I were walking up to the codebase completely unfamiliar with how things are structured, I would grok the need to `init` a mixin class before I would understand the need to decorate all methods using `s3_conn`.

File: shiftmanager/mixins/admin.py (hunk 26 above)
Comment (jklukas) [inline shiftmanager/mixins/admin.py]: #8 is merged now, so we're good here.

File: shiftmanager/redshift.py (hunk 27 above)
Comment (jklukas) [inline shiftmanager/redshift.py]: Looks like the preferred thing here after discussion is to let S3Mixin continue to have a constructor, but call it explicitly here as `S3Mixin.__init__()` rather than using `super`.

**Comment by jklukas** [comment]
Two quick things to address, and I'll be +1 here:

- Call S3Mixin constructor explicitly (https://github.com/SimpleFinance/shiftmanager/pull/7#discussion_r52228230)
- Rebase on master (since #8 was merged)

File: shiftmanager/redshift.py (hunk 27 above)
Comment (wrobstory) [inline shiftmanager/redshift.py]: I think this needs to be `S3Mixin.__init__(self)`

**Comment by xstevens** [comment]
+1

**Comment by wrobstory** [comment]
+1

**Comment by meli-lewis** [comment]
buffy
